store/stat_store.go

```diff
@@ -0,0 +1,442 @@
+
+	validCache bool
+
+	// cachedAverage, cachedMax and cachedPercentiles are the cached derived stats that -
+	// are exposed by the StatStore.
+	cachedAverage     uint64
+	cachedMax         uint64
+	cachedPercentiles []uint64
+}
+
+// tpBucket is a bucket that represents a set of consecutive TimePoints with different
+// timestamps whose values differ less than epsilon.
+// tpBucket essentially represents a time window with a constant value.
+type tpBucket struct {
+	// count is the number of TimePoints represented in the tpBucket.
+	count uint8
+
+	// value is the approximate value of all in the tpBucket, +- epsilon.
+	value uint64
+
+	// max is the maximum value of all TimePoints that have been used to generate the tpBucket.
+	max uint64
+
+	// maxIdx is the number of resolutions after the start time where the max value is located.
+	maxIdx uint8
+}
+
+// putState is a structure that maintains context of the values in the resolution that is currently
+// being inserted.
+// Assumes that Puts are performed in a time-ascending order.
+type putState struct {
+	actualCount uint16
+	average     float64
+	max         uint64
+	stamp       time.Time
+}
+
+func (ss *statStore) Put(tp TimePoint) error {
+	ss.Lock()
+	defer ss.Unlock()
+
+	ss.validCache = false
+
+	// Flatten timestamp to the last multiple of resolution
+	ts := tp.Timestamp.Truncate(ss.resolution)
+
+	lastPutTime := ss.lastPut.stamp
+
+	// Handle the case where the buffer and lastPut are both empty
+	if lastPutTime.Equal(time.Time{}) {
+		ss.resetLastPut(ts, tp.Value.(uint64))
+		return nil
+	}
+
+	if ts.Before(lastPutTime) {
+		// Ignore TimePoints with Timestamps in the past
+		return fmt.Errorf("the provided timepoint has a timestamp in the past")
+	}
+
+	if ts.Equal(lastPutTime) {
+		// update lastPut with the new TimePoint
+		newVal := tp.Value.(uint64)
+		if newVal > ss.lastPut.max {
+			ss.lastPut.max = newVal
+		}
+		oldAvg := ss.lastPut.average
+		n := float64(ss.lastPut.actualCount)
+		ss.lastPut.average = (float64(newVal) + (n * oldAvg)) / (n + 1)
+		ss.lastPut.actualCount++
+		return nil
+	}
+
+	// new point is in the future, lastPut needs to be flushed to the statStore.
+
+	// Determine how many resolutions in the future the new point is at.
+	// The statStore always represents values up until 1 resolution from lastPut.
+	numRes := uint8(0)
+	curr := ts
+	for curr.After(ss.lastPut.stamp) {
+		curr = curr.Add(-ss.resolution)
+		numRes++
+	}
+
+	// Create a new bucket if the buffer is empty
+	if ss.buffer.Front() == nil {
+		ss.newBucket(numRes)
+		ss.resetLastPut(ts, tp.Value.(uint64))
+		for ss.tpCount > ss.windowDuration {
+			ss.rewind()
+		}
+		return nil
+	}
+
+	lastEntry := ss.buffer.Front().Value.(tpBucket)
+	lastAvg := ss.lastPut.average
+
+	// Place lastPut in the latest bucket if the difference from its average
+	// is less than epsilon
+	if uint64(math.Abs(float64(lastEntry.value)-lastAvg)) < ss.epsilon {
+		lastEntry.count += numRes
```

Comment:
Why add numRes here? IIUC, we did not get any data points for the duration between lastPut and current timestamp?